src/powerbox/tools.py

```diff
@@ -175,9 +242,227 @@ def _get_binweights(coords, weights, bins, average=True, bin_ave=True, log_bins=
+    mu : float, optional
+        The minimum value of :math:`\\cos(\theta), \theta = \arctan (k_\\perp/k_\\parallel)`
+        for the sample points generated for the interpolation.
+
+    Returns
+    -------
+    r_n : array-like
+        1D array of radii.
+    phi_n : array-like
+        2D array of azimuthal angles with shape (ndim-1, N), where N is the number of points.
+        phi_n[0,:] :math:`\\in [0,2*\\pi]`, and phi_n[1:,:] :math:`\\in [0,\\pi]`.
+    """
+
+    def generator():
+        r_n, phi_n = regular_angular_generator(
+            bins, dims2avg, angular_resolution=angular_resolution
+        )
+        # sine because the phi_n are wrt x-axis and we need them wrt z-axis.
+        if len(phi_n) == 1:
+            mask = np.sin(phi_n[0, :]) >= mu
+        else:
+            mask = np.all(np.sin(phi_n[1:, :]) >= mu, axis=0)
+        return r_n[mask], phi_n[:, mask]
+
+    return generator()
+
+
+def regular_angular_generator(bins, dims2avg, angular_resolution=0.05):
+    r"""
+    Returns a set of spherical coordinates regularly sampled at a given angular resolution.
+
+    Parameters
+    ----------
+    bins : array-like
+        1D array of radii at which we want to spherically average the field.
+    dims2avg : int
+        The number of dimensions to average over.
+    angular_resolution : float, optional
+        The angular resolution in radians for the sample points for the interpolation.
+        Defaults to 0.1 rad.
+
+    Returns
+    -------
+    r_n : array-like
+        1D array of radii.
+    phi_n : array-like
+        2D array of azimuthal angles with shape (ndim-1, N), where N is the number of points.
+        phi_n[0,:] :math:`\in [0,2*\pi]`, and phi_n[1:,:] :math:`\in [0,\pi]`.
+    """
+
+    def generator():
+        num_angular_bins = np.array(
+            np.max(
+                [
+                    np.round(2 * np.pi * bins / angular_resolution),
+                    np.ones_like(bins) * 100,
+                ],
+                axis=0,
+            ),
+            dtype=int,
+        )
+        phi_i = [np.linspace(0.0, np.pi, n) for n in num_angular_bins]
+        phi_N = [np.linspace(0.0, 2 * np.pi, n) for n in num_angular_bins]
+
+        # Angular resolution is same for all dims
+        phi_n = np.concatenate(
+            [
+                np.array(
+                    np.meshgrid(
+                        *([phi_N[i]] + [phi_i[i]] * (dims2avg - 1)), sparse=False
+                    )
+                ).reshape((dims2avg, num_angular_bins[i] ** dims2avg))
+                for i in range(len(bins))
+            ],
+            axis=-1,
+        )
+        r_n = np.concatenate(
+            [[r] * (num_angular_bins[i] ** dims2avg) for i, r in enumerate(bins)]
+        )
+        return r_n, phi_n
+
+    return generator()
+
+
+def _sample_coords_interpolate(coords, bins, weights, interp_points_generator):
```

Comment:
give the generator a default of None, or indeed, just a default of `regular_angular_generator`